blockchain/chainio.go

```diff
@@ -1403,3 +1432,501 @@ func (b *BlockChain) HeightRange(startHeight, endHeight int32) ([]chainhash.Hash
+// putThresholdCaches function and its associated database transation is closed.
+// This approach is taken to ensure the memory state is not updated until after
+// the atomic database update was successful.
+func (b *BlockChain) markThresholdCachesFlushed() {
+	for i := 0; i < len(b.deploymentCaches); i++ {
+		b.deploymentCaches[i].MarkFlushed()
+	}
+	for i := 0; i < len(b.warningCaches); i++ {
+		b.warningCaches[i].MarkFlushed()
+	}
+}
+
+// dbFetchTresholdCaches uses an existing database transaction to retrieve
+// the threshold state caches from the provided bucket prefix into the given
+// cache parameter.  When the db does not contain any information for a specific
+// id within that cache, that entry will simply be empty.
+func dbFetchTresholdCaches(dbTx database.Tx, caches []thresholdStateCache, bucketPrefix []byte) error {
+	// Nothing to load if the main threshold state caches bucket
+	// doesn't exist.
+	cachesBucket := dbTx.Metadata().Bucket(thresholdBucketName)
+	if cachesBucket == nil {
+		return nil
+	}
+
+	// Loop through each of the cache IDs and load any saved threshold
+	// states.
+	for i := 0; i < len(caches); i++ {
+		// Nothing to do for this cache ID if there is no bucket for it.
+		cacheIDBucketName := thresholdCacheBucket(bucketPrefix, uint32(i))
+		cacheIDBucket := cachesBucket.Bucket(cacheIDBucketName[:])
+		if cacheIDBucket == nil {
+			continue
+		}
+
+		// Load all of the cached block hash to threshold state mappings
+		// from the bucket.
+		err := cacheIDBucket.ForEach(func(k, v []byte) error {
+			// Skip non-hash entries.
+			if len(k) != chainhash.HashSize {
+				return nil
+			}
+
+			var hash chainhash.Hash
+			copy(hash[:], k)
+			caches[i].entries[hash] = ThresholdState(v[0])
+			return nil
+		})
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// invalidateThresholdCaches removes any threshold state caches that are no
+// longer valid.  This can happen if a deployment ID is changed such as when it
+// is reused, or if it is reordered in the parameter definitions.  It is also
+// necessary for specific bits in the warning cache when deployment definitions
+// are added and removed since it could change the expected block versions and
+// hence potentially change the result of the warning states for that bit.
+func (b *BlockChain) invalidateThresholdCaches(cachesBucket database.Bucket) error {
+	deployments := b.chainParams.Deployments[:]
+
+	// Remove any stored deployments that are no longer defined along with
+	// the warning cache associated with their bits.
+	numStoredDeployments, err := dbFetchNumDeployments(cachesBucket)
+	if err != nil {
+		return err
+	}
+	definedDeployments := uint32(len(deployments))
+	for i := definedDeployments; i < numStoredDeployments; i++ {
+		// Nothing to do when nothing is stored for the deployment.
+		deployBucketKey := thresholdCacheBucket(deploymentBucketName, i)
+		deployBucket := cachesBucket.Bucket(deployBucketKey)
```

Comment:
From my reading, and the above comment, it seems that there should be a conditional `continue` here to avoid passing around a possibly `nil` bucket (which may lead to a panic): 
```go
if depolyBucket == nil {
    continue
}
```